File: implementations/cpp/bitsets_and_bitvectors.cpp
#include <cassert>
#include <cstdint>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

namespace dsa {

/**
 * @brief Utilities for single 64-bit machine words.
 */
class WordBits {
public:
    static constexpr size_t BITS_PER_WORD = 64;

    static inline bool test(uint64_t word, size_t bit) {
        return (word & (1ULL << bit)) != 0;
    }

    static inline uint64_t set(uint64_t word, size_t bit) {
        return word | (1ULL << bit);
    }

    static inline uint64_t clear(uint64_t word, size_t bit) {
        return word & ~(1ULL << bit);
    }

    static inline uint64_t flip(uint64_t word, size_t bit) {
        return word ^ (1ULL << bit);
    }

    static inline int popcount(uint64_t word) {
        return __builtin_popcountll(word);
    }

    static inline int ctz(uint64_t word) {
        if (word == 0) return 64;
        return __builtin_ctzll(word);
    }

    static inline uint64_t lowest_set_bit(uint64_t word) {
        return word & (~word + 1ULL);
    }

    template <typename Callback>
    static void iterate_set_bits(uint64_t word, Callback&& cb) {
        while (word != 0) {
            int bit = __builtin_ctzll(word);
            cb(static_cast<size_t>(bit));
            word &= (word - 1ULL);
        }
    }
};

/**
 * @brief Dynamically sized contiguous bitvector backed by 64-bit integer words.
 *
 * Provides O(1) single-bit access and O(N / 64) whole-vector bitwise operations,
 * delivering roughly 64x throughput improvements over boolean arrays.
 */
class BitVector {
private:
    size_t num_bits_;
    std::vector<uint64_t> words_;

    static constexpr size_t BITS_PER_WORD = 64;

    static inline size_t word_index(size_t bit) {
        return bit / BITS_PER_WORD;
    }

    static inline size_t bit_offset(size_t bit) {
        return bit % BITS_PER_WORD;
    }

    static inline uint64_t bit_mask(size_t bit) {
        return 1ULL << bit_offset(bit);
    }

    void mask_trailing_bits() {
        size_t rem = num_bits_ % BITS_PER_WORD;
        if (rem != 0 && !words_.empty()) {
            uint64_t mask = (1ULL << rem) - 1ULL;
            words_.back() &= mask;
        }
    }

public:
    explicit BitVector(size_t num_bits = 0, bool init_val = false)
        : num_bits_(num_bits),
          words_((num_bits + BITS_PER_WORD - 1) / BITS_PER_WORD, init_val ? ~0ULL : 0ULL) {
        mask_trailing_bits();
    }

    size_t size() const noexcept {
        return num_bits_;
    }

    size_t num_words() const noexcept {
        return words_.size();
    }

    bool empty() const noexcept {
        return num_bits_ == 0;
    }

    bool test(size_t i) const {
        if (i >= num_bits_) {
            throw std::out_of_range("BitVector::test index out of range");
        }
        return (words_[word_index(i)] & bit_mask(i)) != 0;
    }

    bool operator[](size_t i) const {
        return test(i);
    }

    void set(size_t i, bool val = true) {
        if (i >= num_bits_) {
            throw std::out_of_range("BitVector::set index out of range");
        }
        if (val) {
            words_[word_index(i)] |= bit_mask(i);
        } else {
            words_[word_index(i)] &= ~bit_mask(i);
        }
    }

    void reset(size_t i) {
        set(i, false);
    }

    void flip(size_t i) {
        if (i >= num_bits_) {
            throw std::out_of_range("BitVector::flip index out of range");
        }
        words_[word_index(i)] ^= bit_mask(i);
    }

    void clear() {
        std::fill(words_.begin(), words_.end(), 0ULL);
    }

    void set_all() {
        std::fill(words_.begin(), words_.end(), ~0ULL);
        mask_trailing_bits();
    }

    size_t count() const {
        size_t total = 0;
        for (uint64_t w : words_) {
            total += static_cast<size_t>(__builtin_popcountll(w));
        }
        return total;
    }

    bool any() const {
        for (uint64_t w : words_) {
            if (w != 0) return true;
        }
        return false;
    }

    bool none() const {
        return !any();
    }

    bool all() const {
        if (num_bits_ == 0) return true;
        return count() == num_bits_;
    }

    std::vector<size_t> get_set_bits() const {
        std::vector<size_t> indices;
        indices.reserve(count());
        for (size_t w_idx = 0; w_idx < words_.size(); ++w_idx) {
            uint64_t word = words_[w_idx];
            size_t base = w_idx * BITS_PER_WORD;
            while (word != 0) {
                int bit = __builtin_ctzll(word);
                indices.push_back(base + static_cast<size_t>(bit));
                word &= (word - 1ULL);
            }
        }
        return indices;
    }

    BitVector operator~() const {
        BitVector result = *this;
        for (size_t i = 0; i < result.words_.size(); ++i) {
            result.words_[i] = ~result.words_[i];
        }
        result.mask_trailing_bits();
        return result;
    }

    BitVector& operator&=(const BitVector& other) {
        if (num_bits_ != other.num_bits_) {
            throw std::invalid_argument("BitVector sizes must match for operator&=");
        }
        for (size_t i = 0; i < words_.size(); ++i) {
            words_[i] &= other.words_[i];
        }
        return *this;
    }

    BitVector& operator|=(const BitVector& other) {
        if (num_bits_ != other.num_bits_) {
            throw std::invalid_argument("BitVector sizes must match for operator|=");
        }
        for (size_t i = 0; i < words_.size(); ++i) {
            words_[i] |= other.words_[i];
        }
        return *this;
    }

    BitVector& operator^=(const BitVector& other) {
        if (num_bits_ != other.num_bits_) {
            throw std::invalid_argument("BitVector sizes must match for operator^=");
        }
        for (size_t i = 0; i < words_.size(); ++i) {
            words_[i] ^= other.words_[i];
        }
        return *this;
    }

    friend BitVector operator&(BitVector lhs, const BitVector& rhs) {
        lhs &= rhs;
        return lhs;
    }

    friend BitVector operator|(BitVector lhs, const BitVector& rhs) {
        lhs |= rhs;
        return lhs;
    }

    friend BitVector operator^(BitVector lhs, const BitVector& rhs) {
        lhs ^= rhs;
        return lhs;
    }

    bool operator==(const BitVector& other) const {
        if (num_bits_ != other.num_bits_) return false;
        return words_ == other.words_;
    }

    bool operator!=(const BitVector& other) const {
        return !(*this == other);
    }

    std::string to_string() const {
        std::string s(num_bits_, '0');
        for (size_t i = 0; i < num_bits_; ++i) {
            if (test(i)) {
                s[i] = '1';
            }
        }
        return s;
    }
};

/**
 * @brief Sieve of Eratosthenes using BitVector visited table.
 * Demonstrates high density and cache efficiency.
 */
inline std::vector<size_t> sieve_primes(size_t limit) {
    if (limit < 2) return {};
    BitVector is_prime(limit + 1, true);
    is_prime.reset(0);
    is_prime.reset(1);

    for (size_t p = 2; p * p <= limit; ++p) {
        if (is_prime.test(p)) {
            for (size_t mult = p * p; mult <= limit; mult += p) {
                is_prime.reset(mult);
            }
        }
    }
    return is_prime.get_set_bits();
}

} // namespace dsa

int main() {
    using namespace dsa;

    std::cout << "[RUNNING] Bitsets and Bitvectors C++17 Verification..." << std::endl;

    // 1. Test WordBits primitives
    {
        uint64_t w = 0;
        w = WordBits::set(w, 3);
        w = WordBits::set(w, 7);
        w = WordBits::set(w, 63);
        assert(WordBits::test(w, 3));
        assert(WordBits::test(w, 7));
        assert(WordBits::test(w, 63));
        assert(!WordBits::test(w, 0));
        assert(!WordBits::test(w, 62));

        assert(WordBits::popcount(w) == 3);
        assert(WordBits::ctz(w) == 3);
        assert(WordBits::lowest_set_bit(w) == (1ULL << 3));

        std::vector<size_t> collected;
        WordBits::iterate_set_bits(w, [&](size_t bit) { collected.push_back(bit); });
        assert(collected.size() == 3);
        assert(collected[0] == 3 && collected[1] == 7 && collected[2] == 63);

        w = WordBits::clear(w, 7);
        assert(!WordBits::test(w, 7));
        assert(WordBits::popcount(w) == 2);

        w = WordBits::flip(w, 0);
        assert(WordBits::test(w, 0));
        w = WordBits::flip(w, 0);
        assert(!WordBits::test(w, 0));
    }

    // 2. Test BitVector initialization and boundary sizes
    {
        BitVector b0(0);
        assert(b0.size() == 0);
        assert(b0.num_words() == 0);
        assert(b0.empty());
        assert(b0.count() == 0);
        assert(b0.all());

        BitVector b1(1, true);
        assert(b1.size() == 1);
        assert(b1.num_words() == 1);
        assert(b1.test(0));
        assert(b1.count() == 1);
        assert(b1.all());

        BitVector b64(64, true);
        assert(b64.size() == 64);
        assert(b64.num_words() == 1);
        assert(b64.count() == 64);
        assert(b64.all());

        BitVector b65(65, false);
        assert(b65.size() == 65);
        assert(b65.num_words() == 2);
        assert(b65.count() == 0);
        assert(b65.none());
    }

    // 3. Test BitVector single-bit modifications across multiple words
    {
        BitVector bv(200, false);
        assert(bv.size() == 200);
        assert(bv.num_words() == 4);

        bv.set(0);
        bv.set(63);
        bv.set(64);
        bv.set(127);
        bv.set(128);
        bv.set(199);

        assert(bv.test(0));
        assert(bv.test(63));
        assert(bv.test(64));
        assert(bv.test(127));
        assert(bv.test(128));
        assert(bv.test(199));
        assert(!bv.test(1));
        assert(!bv.test(65));
        assert(!bv.test(198));

        assert(bv.count() == 6);
        assert(bv.any());
        assert(!bv.none());
        assert(!bv.all());

        auto set_bits = bv.get_set_bits();
        assert(set_bits.size() == 6);
        assert((set_bits == std::vector<size_t>{0, 63, 64, 127, 128, 199}));

        bv.flip(64);
        assert(!bv.test(64));
        assert(bv.count() == 5);

        bv.flip(64);
        assert(bv.test(64));
        assert(bv.count() == 6);
    }

    // 4. Test Whole-Vector Bitwise Operations and Trailing Bit Masking
    {
        BitVector a(70, false);
        BitVector b(70, false);

        a.set(10);
        a.set(65);
        a.set(69);

        b.set(10);
        b.set(20);
        b.set(65);

        BitVector and_res = a & b;
        assert(and_res.count() == 2);
        assert(and_res.test(10) && and_res.test(65));
        assert(!and_res.test(20) && !and_res.test(69));

        BitVector or_res = a | b;
        assert(or_res.count() == 4);
        assert(or_res.test(10) && or_res.test(20) && or_res.test(65) && or_res.test(69));

        BitVector xor_res = a ^ b;
        assert(xor_res.count() == 2);
        assert(xor_res.test(20) && xor_res.test(69));
        assert(!xor_res.test(10) && !xor_res.test(65));

        BitVector not_a = ~a;
        assert(not_a.size() == 70);
        assert(not_a.count() == 70 - 3);
        assert(!not_a.test(10) && !not_a.test(65) && !not_a.test(69));
        assert(not_a.test(0) && not_a.test(68));

        // Test set_all and clear
        a.set_all();
        assert(a.count() == 70);
        assert(a.all());
        a.clear();
        assert(a.count() == 0);
        assert(a.none());
    }

    // 5. Test Exceptions on Out of Range & Mismatched Sizes
    {
        BitVector bv(10);
        bool caught = false;
        try {
            bv.test(10);
        } catch (const std::out_of_range&) {
            caught = true;
        }
        assert(caught);

        caught = false;
        try {
            bv.set(10);
        } catch (const std::out_of_range&) {
            caught = true;
        }
        assert(caught);

        BitVector other(11);
        caught = false;
        try {
            bv &= other;
        } catch (const std::invalid_argument&) {
            caught = true;
        }
        assert(caught);
    }

    // 6. Test Prime Sieve with BitVector
    {
        auto primes = sieve_primes(30);
        std::vector<size_t> expected = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
        assert(primes == expected);

        auto primes100 = sieve_primes(100);
        assert(primes100.size() == 25);
    }

    std::cout << "[PASSED] Bitsets and Bitvectors C++17 All Tests Passed!" << std::endl;
    return 0;
}

File: implementations/python/bitsets_and_bitvectors.py
"""
Bitsets and Bitvectors implementation in Python.

Provides:
1. Word-level bit manipulation helper functions.
2. Dynamically sized BitVector backed by 64-bit word chunks.
3. Word-level bitwise operations (&, |, ^, ~).
4. Sieve of Eratosthenes prime generator using BitVector.
5. Comprehensive unit tests.
"""

import unittest
from typing import List, Callable


class WordBits:
    """Utilities for 64-bit integer words."""

    BITS_PER_WORD = 64
    WORD_MASK = (1 << 64) - 1

    @staticmethod
    def test(word: int, bit: int) -> bool:
        return (word & (1 << bit)) != 0

    @staticmethod
    def set(word: int, bit: int) -> int:
        return word | (1 << bit)

    @staticmethod
    def clear(word: int, bit: int) -> int:
        return word & ~(1 << bit)

    @staticmethod
    def flip(word: int, bit: int) -> int:
        return word ^ (1 << bit)

    @staticmethod
    def popcount(word: int) -> int:
        return word.bit_count()

    @staticmethod
    def ctz(word: int) -> int:
        """Count trailing zeros in 64-bit word."""
        if word == 0:
            return 64
        lsb = word & -word
        return lsb.bit_length() - 1

    @staticmethod
    def lowest_set_bit(word: int) -> int:
        return word & -word

    @staticmethod
    def iterate_set_bits(word: int, callback: Callable[[int], None]) -> None:
        while word != 0:
            lsb = word & -word
            bit = lsb.bit_length() - 1
            callback(bit)
            word &= (word - 1)


class BitVector:
    """
    Dynamically sized bitvector backed by 64-bit integer words.
    """

    BITS_PER_WORD = 64
    WORD_MASK = (1 << 64) - 1

    def __init__(self, num_bits: int = 0, init_val: bool = False):
        if num_bits < 0:
            raise ValueError("num_bits must be non-negative")
        self._num_bits = num_bits
        num_words = (num_bits + self.BITS_PER_WORD - 1) // self.BITS_PER_WORD
        init_word = self.WORD_MASK if init_val else 0
        self._words = [init_word] * num_words
        self._mask_trailing_bits()

    def _mask_trailing_bits(self) -> None:
        rem = self._num_bits % self.BITS_PER_WORD
        if rem != 0 and self._words:
            mask = (1 << rem) - 1
            self._words[-1] &= mask

    def __len__(self) -> int:
        return self._num_bits

    @property
    def num_words(self) -> int:
        return len(self._words)

    def _check_index(self, i: int) -> None:
        if i < 0 or i >= self._num_bits:
            raise IndexError(f"BitVector index {i} out of range for size {self._num_bits}")

    def test(self, i: int) -> bool:
        self._check_index(i)
        w_idx = i // self.BITS_PER_WORD
        offset = i % self.BITS_PER_WORD
        return (self._words[w_idx] & (1 << offset)) != 0

    def __getitem__(self, i: int) -> bool:
        return self.test(i)

    def set(self, i: int, val: bool = True) -> None:
        self._check_index(i)
        w_idx = i // self.BITS_PER_WORD
        offset = i % self.BITS_PER_WORD
        if val:
            self._words[w_idx] |= (1 << offset)
        else:
            self._words[w_idx] &= ~(1 << offset)

    def reset(self, i: int) -> None:
        self.set(i, False)

    def flip(self, i: int) -> None:
        self._check_index(i)
        w_idx = i // self.BITS_PER_WORD
        offset = i % self.BITS_PER_WORD
        self._words[w_idx] ^= (1 << offset)

    def clear(self) -> None:
        self._words = [0] * len(self._words)

    def set_all(self) -> None:
        self._words = [self.WORD_MASK] * len(self._words)
        self._mask_trailing_bits()

    def count(self) -> int:
        return sum(w.bit_count() for w in self._words)

    def any(self) -> bool:
        return any(w != 0 for w in self._words)

    def none(self) -> bool:
        return not self.any()

    def all(self) -> bool:
        if self._num_bits == 0:
            return True
        return self.count() == self._num_bits

    def get_set_bits(self) -> List[int]:
        res = []
        for w_idx, word in enumerate(self._words):
            base = w_idx * self.BITS_PER_WORD
            while word != 0:
                lsb = word & -word
                bit = lsb.bit_length() - 1
                res.append(base + bit)
                word &= (word - 1)
        return res

    def __invert__(self) -> 'BitVector':
        res = BitVector(self._num_bits)
        res._words = [w ^ self.WORD_MASK for w in self._words]
        res._mask_trailing_bits()
        return res

    def __and__(self, other: 'BitVector') -> 'BitVector':
        if self._num_bits != other._num_bits:
            raise ValueError("BitVector sizes must match for bitwise AND")
        res = BitVector(self._num_bits)
        res._words = [w1 & w2 for w1, w2 in zip(self._words, other._words)]
        return res

    def __or__(self, other: 'BitVector') -> 'BitVector':
        if self._num_bits != other._num_bits:
            raise ValueError("BitVector sizes must match for bitwise OR")
        res = BitVector(self._num_bits)
        res._words = [w1 | w2 for w1, w2 in zip(self._words, other._words)]
        return res

    def __xor__(self, other: 'BitVector') -> 'BitVector':
        if self._num_bits != other._num_bits:
            raise ValueError("BitVector sizes must match for bitwise XOR")
        res = BitVector(self._num_bits)
        res._words = [w1 ^ w2 for w1, w2 in zip(self._words, other._words)]
        return res

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, BitVector):
            return False
        return self._num_bits == other._num_bits and self._words == other._words

    def to_string(self) -> str:
        return "".join("1" if self.test(i) else "0" for i in range(self._num_bits))


def sieve_primes(limit: int) -> List[int]:
    """Compute prime numbers up to limit using BitVector visited table."""
    if limit < 2:
        return []
    is_prime = BitVector(limit + 1, True)
    is_prime.reset(0)
    is_prime.reset(1)

    p = 2
    while p * p <= limit:
        if is_prime.test(p):
            for mult in range(p * p, limit + 1, p):
                is_prime.reset(mult)
        p += 1
    return is_prime.get_set_bits()


class TestBitsetsAndBitvectors(unittest.TestCase):
    def test_word_bits(self):
        w = 0
        w = WordBits.set(w, 5)
        w = WordBits.set(w, 63)
        self.assertTrue(WordBits.test(w, 5))
        self.assertTrue(WordBits.test(w, 63))
        self.assertFalse(WordBits.test(w, 0))
        self.assertEqual(WordBits.popcount(w), 2)
        self.assertEqual(WordBits.ctz(w), 5)
        self.assertEqual(WordBits.lowest_set_bit(w), 1 << 5)

        collected = []
        WordBits.iterate_set_bits(w, collected.append)
        self.assertEqual(collected, [5, 63])

        w = WordBits.clear(w, 5)
        self.assertFalse(WordBits.test(w, 5))
        self.assertEqual(WordBits.popcount(w), 1)

        w = WordBits.flip(w, 10)
        self.assertTrue(WordBits.test(w, 10))
        w = WordBits.flip(w, 10)
        self.assertFalse(WordBits.test(w, 10))

    def test_bitvector_sizes_and_init(self):
        b0 = BitVector(0)
        self.assertEqual(len(b0), 0)
        self.assertEqual(b0.num_words, 0)
        self.assertEqual(b0.count(), 0)
        self.assertTrue(b0.all())

        b1 = BitVector(1, True)
        self.assertEqual(len(b1), 1)
        self.assertEqual(b1.count(), 1)
        self.assertTrue(b1.test(0))

        b64 = BitVector(64, True)
        self.assertEqual(len(b64), 64)
        self.assertEqual(b64.num_words, 1)
        self.assertEqual(b64.count(), 64)

        b65 = BitVector(65, False)
        self.assertEqual(len(b65), 65)
        self.assertEqual(b65.num_words, 2)
        self.assertEqual(b65.count(), 0)
        self.assertTrue(b65.none())

    def test_modifications(self):
        bv = BitVector(150, False)
        indices = [0, 63, 64, 100, 149]
        for idx in indices:
            bv.set(idx)

        for idx in range(150):
            self.assertEqual(bv.test(idx), idx in indices)

        self.assertEqual(bv.count(), 5)
        self.assertTrue(bv.any())
        self.assertFalse(bv.none())
        self.assertFalse(bv.all())
        self.assertEqual(bv.get_set_bits(), indices)

        bv.flip(64)
        self.assertFalse(bv.test(64))
        self.assertEqual(bv.count(), 4)

    def test_bitwise_operations(self):
        a = BitVector(70, False)
        b = BitVector(70, False)

        a.set(10)
        a.set(65)
        a.set(69)

        b.set(10)
        b.set(20)
        b.set(65)

        and_res = a & b
        self.assertEqual(and_res.count(), 2)
        self.assertTrue(and_res.test(10) and and_res.test(65))
        self.assertFalse(and_res.test(20) or and_res.test(69))

        or_res = a | b
        self.assertEqual(or_res.count(), 4)

        xor_res = a ^ b
        self.assertEqual(xor_res.count(), 2)
        self.assertTrue(xor_res.test(20) and xor_res.test(69))

        not_a = ~a
        self.assertEqual(len(not_a), 70)
        self.assertEqual(not_a.count(), 70 - 3)
        self.assertFalse(not_a.test(10))
        self.assertTrue(not_a.test(0))

    def test_sieve(self):
        primes = sieve_primes(30)
        self.assertEqual(primes, [2, 3, 5, 7, 11, 13, 17, 19, 23, 29])
        primes100 = sieve_primes(100)
        self.assertEqual(len(primes100), 25)

    def test_exceptions(self):
        bv = BitVector(10)
        with self.assertRaises(IndexError):
            bv.test(10)
        with self.assertRaises(IndexError):
            bv.set(15)
        with self.assertRaises(ValueError):
            _ = bv & BitVector(11)


if __name__ == '__main__':
    unittest.main()
